src/sentry/sentry_metrics/querying/visitors/query_expression.py

```diff
@@ -258,3 +266,71 @@ def _group_bys_as_string(self, group_bys: list[Column | AliasedExpression] | Non
                 string_group_bys.add(group_by.name)
 
         return string_group_bys
+
+
+class UnitsNormalizationVisitor(QueryExpressionVisitor[QueryExpression]):
+    """
+    Visitor that recursively transforms the `QueryExpression` components to have the same unit. Throws an error in
+    case units are incompatible.
+    """
+
+    UNITLESS_AGGREGATES = {"count", "count_unique"}
+
+    def __init__(self):
+        self._unit_family = None
+        self._reference_unit = None
+        self._scaling_factor = None
+
+        self._is_formula = False
+
+    def _visit_formula(self, formula: Formula) -> QueryExpression:
+        self._is_formula = True
+        return super()._visit_formula(formula)
+
+    def _visit_timeseries(self, timeseries: Timeseries) -> QueryExpression:
+        extracted_unit = self._extract_unit(timeseries=timeseries)
+        if extracted_unit is not None:
+            unit_family, reference_unit, unit = get_unit_family_and_unit(extracted_unit)
+            # If we encounter multiple unit families in a `QueryExpression`, we want to unwind and not apply any
+            # units normalization.
+            if self._unit_family is not None and unit_family != self._unit_family:
+                raise NonNormalizableUnitsError("Multiple unit families are found in the formula")
+
+            # We set the first seen unit family, irrespectively if a unit is found, since if it's not found, the family
+            # will be unknown.
+            self._unit_family = unit_family
+
+            if reference_unit is not None and unit is not None:
+                self._reference_unit = reference_unit
+                self._scaling_factor = unit.scaling_factor
+                return unit.apply_on_timeseries(timeseries)
+
+        return timeseries
+
+    def _extract_unit(self, timeseries: Timeseries) -> str | None:
+        # If the aggregate doesn't support unit normalization, we will skip it.
+        if timeseries.aggregate in self.UNITLESS_AGGREGATES:
+            raise NonNormalizableUnitsError(
+                f"The aggregate {timeseries.aggregate} doesn't need unit normalization"
+            )
+
+        parsed_mri = parse_mri(timeseries.metric.mri)
+        if parsed_mri is not None:
+            return parsed_mri.unit
+
+        raise NonNormalizableUnitsError(
+            "Units normalization can't be run if not all components have a metric mri"
+        )
+
+    def get_units_metadata(
```

Comment:
Not the best solution, since it exposes a method that returns different data based on the call of the `visit` method, but for the sake of simplicity I opted to implement it this way.